utils/containers/lock_free/lf_queue.c.tpl

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include "utils/containers/lock_free/lf_queue_{TYPE}.h"
+#include <stdlib.h>
+#include <string.h>
+
+void iota_lock_free_queue_umm_init_owner(iota_lf_umm_queue_{TYPE}_t* const queue,
+                                         uint32_t element_size) {
+  lfds711_queue_umm_init_valid_on_current_logical_core(
+      &queue->queue, &queue->queue_element_dummy, NULL);
+
+  lfds711_freelist_init_valid_on_current_logical_core(&queue->freelist, NULL, 0,
+                                                      NULL);
+}
+
+void iota_lock_free_queue_umm_init_user(iota_lf_umm_queue_{TYPE}_t* const queue) {
+  LFDS711_MISC_MAKE_VALID_ON_CURRENT_LOGICAL_CORE_INITS_COMPLETED_BEFORE_NOW_ON_ANY_OTHER_LOGICAL_CORE;
+}
+
+void* iota_lock_free_queue_umm_free(iota_lf_umm_queue_{TYPE}_t* const queue) {
+  if (queue == NULL) {
+    return;
+  }
+  lfds711_queue_umm_cleanup(&queue->queue, NULL);
+
+  struct lfds711_freelist_element* fe;
+  iota_lf_queue_umm_{TYPE}_t* queue_element;
+  while (lfds711_freelist_pop(&queue->freelist, &fe, NULL)) {
+    queue_element = LFDS711_FREELIST_GET_VALUE_FROM_ELEMENT(*fe);
+    free(queue_element);
+  }
+  lfds711_freelist_cleanup(&queue->freelist, NULL);
+}
+
+retcode_t iota_lock_free_queue_umm_enqueue(iota_lf_umm_queue_{TYPE}_t* const queue,
+                                           void const* const data) {
+  iota_lf_queue_umm_{TYPE}_t* p = NULL;
+
+  if (queue == NULL || data == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  struct lfds711_freelist_element* fe;
+  if (lfds711_freelist_pop(&queue->freelist, &fe, NULL)) {
+    p = LFDS711_FREELIST_GET_VALUE_FROM_ELEMENT(*fe);
+  } else if ((p = malloc(sizeof(iota_lf_queue_umm_{TYPE}_t))) == NULL) {
+    return RC_OOM;
+  }
+
+  memcpy(&p->data, data, sizeof(p->data));
+  LFDS711_QUEUE_UMM_SET_VALUE_IN_ELEMENT(p->queue_element, p);
```

Comment:
+ `LFDS711_FREELIST_SET_VALUE_IN_ELEMENT(p->freelist_element, p);`

Reply:
also probably should only happen if we had to malloc

Reply:
yes